pkg/queues/amqp_provider.go

```diff
@@ -0,0 +1,190 @@
+	url     string
+	log     logrus.FieldLogger
+	wg      *sync.WaitGroup
+	queues  []*amqpQueue
+	stopped atomic.Bool
+}
+
+func NewAmqpProvider(url string, log logrus.FieldLogger) Provider {
+	var wg sync.WaitGroup
+	wg.Add(1)
+	return &amqpProvider{
+		doneCh: make(chan struct{}),
+		url:    url,
+		log:    log,
+		wg:     &wg,
+	}
+}
+
+func (r *amqpProvider) newQueue(queueName string) (*amqpQueue, error) {
+	var (
+		err        error
+		connection *amqp.Connection
+		channel    *amqp.Channel
+	)
+	if r.stopped.Load() {
+		return nil, errors.New("provider is stopped")
+	}
+	defer func() {
+		if err != nil {
+			if channel != nil {
+				channel.Close()
+			}
+			if connection != nil {
+				connection.Close()
+			}
+		}
+	}()
+	connection, err = amqp.Dial(r.url)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create connection: %w", err)
+	}
+	channel, err = connection.Channel()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create channel: %w", err)
+	}
+	_, err = channel.QueueDeclare(queueName,
+		true,  // durable
+		false, // auto delete
+		false, // exclusive
+		false, // no wait
+		nil)   // args
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to declare queue %s: %w", queueName, err)
+	}
+	ret := &amqpQueue{
+		name:       queueName,
+		connection: connection,
+		channel:    channel,
+		log:        r.log,
+		doneCh:     r.doneCh,
+		wg:         r.wg,
+	}
+	r.queues = append(r.queues, ret)
+	return ret, nil
+}
+
+func (a *amqpProvider) NewConsumer(queueName string) (Consumer, error) {
+	return a.newQueue(queueName)
+}
+
+func (a *amqpProvider) NewPublisher(queueName string) (Publisher, error) {
+	return a.newQueue(queueName)
+}
+
+func (a *amqpProvider) Stop() {
+	if a.stopped.Swap(true) {
+		return
+	}
+	defer a.wg.Done()
+	close(a.doneCh)
+	for _, q := range a.queues {
+		q.Close()
+	}
+}
+
+func (a *amqpProvider) Wait() {
+	a.wg.Wait()
+}
+
+type amqpQueue struct {
+	doneCh     chan struct{}
+	connection *amqp.Connection
+	channel    *amqp.Channel
+	name       string
+	wg         *sync.WaitGroup
+	log        logrus.FieldLogger
+	closed     atomic.Bool
+}
+
+func (r *amqpQueue) Publish(payload []byte) error {
+	if r.closed.Load() {
+		return errors.New("queue is closed")
+	}
+	return r.channel.Publish("",
+		r.name, // queue name
+		false,  // mandatory
+		false,  // immediate
+		amqp.Publishing{
+			DeliveryMode: amqp.Persistent,
+			ContentType:  "text/plain",
+			Body:         payload,
+		})
+}
+
+func (r *amqpQueue) Consume(ctx context.Context, handler ConsumeHandler) error {
+	ctxWithCancel, cancel := context.WithCancel(ctx)
+	msgs, err := r.channel.ConsumeWithContext(ctxWithCancel,
+		r.name,
+		"",    // consumer (exchange)
+		false, // auto ack
+		false, // exclusive
+		false, // no local
+		false, // no wait
+		nil)   // args
+	if err != nil {
+		cancel()
+		return err
+	}
+	r.wg.Add(1)
+	go func() {
+		var err error
+		defer r.wg.Done()
+		defer cancel()
+		for {
+			select {
+			case <-r.doneCh:
```

Comment:
Since you have ctx as long as the parent ctx CancelFn is on the struct you don't need the done channel. When the parent is canceled so will its childern.

```suggestion
			case <-ctx.Done():
```

Reply:
It wouldn't help if the provider is stopped, since now the ctx is received as parameter.